### DISCUSSION TITLE
AttributeError: type object 'Language' has no attribute 'factory'

### DISCUSSION THREAD

**Discussion by ceteri** [explosion/spaCy#7435] [Help: Coding & Implementations] [answered]
I'm feeling like I'm going crazy just trying to find any difference between the following two cases :)  Any help or advice is much appreciated!

## How to reproduce the behaviour
When running the following code within a JupyterLab notebook:
```
import spacy
from spacy.language import Language

class MyComponent:
    def __init__(self, some_setting):
        print("got here")

@Language.factory("my_component", default_config={"some_setting": True})
def my_component(nlp, name, some_setting: bool):
    return MyComponent(some_setting=some_setting)

nlp = spacy.blank("en")
nlp.add_pipe("my_component", config={"some_setting": False})
```

It raises `AttributeError: type object 'Language' has no attribute 'factory'`:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-7d9b9734bde1> in <module>
      6         print("got here")
      7 
----> 8 @Language.factory("my_component", default_config={"some_setting": True})
      9 def my_component(nlp, name, some_setting: bool):
     10     return MyComponent(some_setting=some_setting)

AttributeError: type object 'Language' has no attribute 'factory'
```

This same code runs fine when run as a Python script within the same environment.  Our unit tests didn't identify any problem with the code when we migrated pytextrank to spaCy 3.x by using a custom pipeline component factory. 

However, we've had at least one user report a bug, and now I'm getting this when I try to run code samples in Jupyter too.

Has anyone else been seeing this `AttributeError` show up?  The last similar mention that I could find was <https://github.com/explosion/spaCy/issues/1695>

## Your Environment
* Operating System: macOS 10.13.6 (17G14042)
- **spaCy version:** 3.0.5
- **Platform:** Darwin-17.7.0-x86_64-i386-64bit
- **Python version:** 3.7.9
- **Pipelines:** en_core_web_sm (3.0.0)

Many thanks

**Comment by svlandeg** [comment]
Hmmm, that's weird. I don't suppose there could be another `Language` class definition obscuring things here? What if you import spaCy's `Language` under a different name and use that?

**Comment by ceteri** [comment]
Great idea, thanks @svlandeg !

I tested that in this notebook:

- https://github.com/DerwenAI/pytextrank/blob/main/wip/error.ipynb
- https://github.com/DerwenAI/pytextrank/issues/99#issuecomment-799691604

It turns out that the `spacy.language.Language` class appears to be defined in three different library modules:
```
'spacy.lang.xx', 'Language', spacy.language.Language
'spacy.language', 'Language', spacy.language.Language
'spacy.pipeline.pipes', 'Language', spacy.language.Language
```

The notebook also uses `hasattr()` to test for the `factory` attribute, which doesn't appear to be in any of those three.

Since this uses the `inspect` library, are there other tests that we could do on these three `Language` class definitions, perhaps to help isolate the issue?

**Reply by svlandeg** [reply]
Ah, but wait! `spacy.pipeline.pipes` refers to a Cython class that has been refactored and removed in v3. So perhaps your environment still accesses that somehow - could you try cleaning up those files manually? 

`spacy.lang.xx` really should only refer to the same `Language` object in `spacy.language`, which definitely does have a `factory` method...

**Comment by svlandeg** [comment, marked as answer]
Actually ... are you super sure your notebook is running the correct virtual environment with spaCy v3?

For instance, `spacy._ml` doesn't exist anymore either (but used to, in v2)

**Reply by ceteri** [reply]
That's what I was thinking initially, that somehow JupyterLab must not be launching within the correct virtual environment with spaCy v3. 

However, I'm running this code as a Python script (runs fine) and then launching JupyterLab to run this notebook (raises exception) both from the same environment. This environment has the version info listed above.

My steps to reproduce this, starting from scratch:

1. `cd ~/src/tmp`
2. `git clone https://github.com/DerwenAI/pytextrank.git`
3. `cd pytextrank`
4. `python3 -m venv venv`
5.  `source venv/bin/activate`
6. `pip install -r requirements.txt`
7. `python -m spacy download en_core_web_sm`

then:

```
(venv) (base) derwen:~/src/tmp/pytextrank$ env | grep VIRTUAL_ENV 
VIRTUAL_ENV=/Users/paco/src/tmp/pytextrank/venv
(venv) (base) derwen:~/src/tmp/pytextrank$ pip freeze | grep spacy
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0-py3-none-any.whl
spacy==3.0.5
spacy-legacy==3.0.1
(venv) (base) derwen:~/src/tmp/pytextrank$ jupyter-lab
[I 14:54:17.134 LabApp] JupyterLab extension loaded from /opt/anaconda3/lib/python3.7/site-packages/jupyterlab
[I 14:54:17.134 LabApp] JupyterLab application directory is /opt/anaconda3/share/jupyter/lab
[I 14:54:17.137 LabApp] Serving notebooks from local directory: /Users/paco/src/tmp/pytextrank
[I 14:54:17.137 LabApp] The Jupyter Notebook is running at:
[I 14:54:17.138 LabApp] http://localhost:8888/?token=5e80cef439efda32a49891a0d93ebb2952d5809a629212ef
[I 14:54:17.138 LabApp]  or http://127.0.0.1:8888/?token=5e80cef439efda32a49891a0d93ebb2952d5809a629212ef
[I 14:54:17.138 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 14:54:17.144 LabApp] 
    
    To access the notebook, open this file in a browser:
        file:///Users/paco/Library/Jupyter/runtime/nbserver-32298-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=5e80cef439efda32a49891a0d93ebb2952d5809a629212ef
     or http://127.0.0.1:8888/?token=5e80cef439efda32a49891a0d93ebb2952d5809a629212ef
[I 14:54:19.365 LabApp] 301 GET /lab/workspaces/auto-2/?clone (127.0.0.1) 1.90ms
[I 14:54:21.126 LabApp] Build is up to date
[I 14:54:24.266 LabApp] Kernel started: 2242b31b-e53e-4956-b929-f82fcde520ee
[I 14:54:32.975 LabApp] Kernel restarted: 2242b31b-e53e-4956-b929-f82fcde520ee
```

Trying to step through this logically, one difference is JupyterLab.  Is there any way possible that JupyterLab is maintaining some kind of state from a previous environment, perhaps blending in some paths or other environment variables?  Something that could affect its Py 3 kernel env?

Should I try to uninstall explicitly for any of those spaCy v2 modules?

FWIW, I have two laptops, a macOS PowerBook Pro and a KFocus Ubuntu laptop.  This issue only happens on macOS for me; cannot reproduce it on Ubuntu.

Also, many thanks for your help. 
I feel embarrassed about having such a strange case here.

**Reply by svlandeg** [reply]
Hm. I don't suppose you could try with an entirely new virtual environment? If you never installed v2 in that venv, you shouldn't see any references to `_ml` or `pipes.pyx` anymore...

**Comment by ceteri** [comment]
Thank you @svlandeg for all of your help here.

This was resolved by moving to different file paths, new repo checkout, new environment, etc., and finally then I tried uninstalling spaCy to see what was happening. It was a problem with Jupyter and also the Python environment that laptop. Using `python` at the command line creates a Py environment as described above by version number.  So does use of `python3` with the same version. I should've been checking the binary path since those aren't the same.

Somehow now Jupyter is calling a different `python` executable, which I found was running spaCy 2.3.5 -- even when launched within the `venv` virtual environment that has spaCy 3 instead.

Two takeaways for me:

1. Need to be able to configure my JupyterLab set up so that it runs a specific Python executable and environment
2. Need to find the Python executable **path** in addition to its version when troubleshooting an issue

This latter point might perhaps also be helpful for the the spaCy instructions about how to generate markdown for "Your Environment" too?

**Reply by svlandeg** [reply]
Oh wow, OK, thanks for reporting back, and happy to hear you got it sorted out!

**Comment by ceteri** [comment]
For people reporting issues related to Jupyter and other notebook environments, it might be helpful to run:
```
import sys
print(sys.executable)
```

Since JupyterLab is a library, although it has the ability to run as a CLI also, it matters where the installed library lives.

For example, in my case Conda had an old environment, with JupyterLab installed, but spaCy 2 since I don't use Conda generally.  

While I had installed JupyterLab and spaCy 3 via pip into my virtual environment, the shell's path picked up the executable for JupyterLab from the Conda directory, not from my `venv` that I was running.

The giveaway was the I couldn't reproduce the error case above in Jupyter, only in JupyterLab. Also, I saw other libraries-acting-as-CLI hit this error case as well, such as `pylint`.

IMO, the installer for Conda seems relatively aggressive in terms of how it rewrites a bash profile.  Also, the docs for JupyterLab could stand some more love.  Speaking as former co-chair for JupyterCon, now there's a project to tackle :)

This person ran into a related issue:
<https://stackoverflow.com/questions/53196908/jupyter-lab-incorrectly-uses-base-python-executable-jupyter-notebook-correctly>

**Comment by Warlock1993** [comment]
@ceteri  ....  I suggest to copy the path for the command "python -m spacy info" into the Path of environmental variables. It solved the issue in Jupyter Notebook for me
